internal/evmreader/claim.go

```diff
@@ -0,0 +1,220 @@
+				"most recent block", mostRecentBlockNumber)
+
+			r.readAndUpdateClaims(ctx, apps, lastClaimCheck, mostRecentBlockNumber)
+
+		} else if mostRecentBlockNumber < lastClaimCheck {
+			slog.Warn(
+				"Not reading claim acceptance: most recent block is lower than the last processed one", //nolint:lll
+				"apps", appAddresses,
+				"last claim check block", lastClaimCheck,
+				"most recent block", mostRecentBlockNumber,
+			)
+		} else {
+			slog.Info("Not reading claim acceptance: already checked the most recent blocks",
+				"apps", appAddresses,
+				"last claim check block", lastClaimCheck,
+				"most recent block", mostRecentBlockNumber,
+			)
+		}
+
+	}
+}
+
+func (r *EvmReader) readAndUpdateClaims(
+	ctx context.Context,
+	apps []application,
+	lastClaimCheck, mostRecentBlockNumber uint64,
+) {
+
+	// DISCLAIMER: The current algorithm will only handle Authority.
+	// To handle Quorum, node needs to handle acceptance events
+	// that can happen before claim submission
+
+	// Classify them by same IConsensusAddress
+	sameConsensusApps := indexApps(keyByIConsensus, apps)
+	for iConsensusAddress, apps := range sameConsensusApps {
+
+		appAddresses := appToAddresses(apps)
+
+		// All these apps shares the same IConsensus
+		if len(apps) == 0 {
+			continue
+		}
+		consensusContract := apps[0].consensusContract
+
+		// Retrieve Claim Acceptance Events from blockchain
+		appClaimAcceptanceEventMap, err := r.readClaimAcceptanceFromBlockchain(
+			ctx, consensusContract, appAddresses, lastClaimCheck+1, mostRecentBlockNumber)
+		if err != nil {
+			slog.Error("Error reading claim acceptance status",
+				"apps", apps,
+				"IConsensus", iConsensusAddress,
+				"start", lastClaimCheck,
+				"end", mostRecentBlockNumber,
+				"error", err)
+			continue
+		}
+
+		// Check events against Epochs
+		for app, claimAcceptances := range appClaimAcceptanceEventMap {
+
+			epochs := []*Epoch{}
+			for _, claimAcceptance := range claimAcceptances {
+
+				// Get Previous claims and update their statuses
+				// rejecting all
+				previousClaims, err := r.repository.GetPreviousSubmittedClaims(
+					ctx, app, claimAcceptance.LastProcessedBlockNumber.Uint64())
+				if err != nil {
+					slog.Error("Error retrieving previous submitted claims",
+						"app", app, "error", err)
+				}
+
+				for _, previousClaim := range previousClaims {
+					previousClaim.Status = EpochStatusClaimRejected
+					epochs = append(epochs, &previousClaim)
+					slog.Warn("Claim rejected",
+						"app", app,
+						"lastBlock", previousClaim.LastBlock,
+						"hash", previousClaim.ClaimHash)
+				}
+
+				// Get  Claim
+				claim, err := r.repository.GetEpoch(
+					ctx, calculateEpochIndex(
+						r.epochLengthCache[app],
+						claimAcceptance.LastProcessedBlockNumber.Uint64()),
+					app)
+				if err != nil {
+					slog.Error("Error retrieving claim", "app", app, "error", err)
+				}
+
+				// Check Claim
+				if claim == nil {
+					slog.Error("Got unknown claim event",
```

Comment:
This claim could be nil because the repository failed to retrieve the epoch due to a network issue, for example. Then your error reason would be incorrect.

Reply:
The error before was not correctly handled. Corrected it. The only way the claim is nil, is being absent from the DB